### PR TITLE
feat(opentelemetry): Add capability to atmo to use opentel when in proxy mode

### LINKED ISSUE #143: Add in OpenTelemetry to atmo proxy
This is a sibling issue to https://github.com/suborbital/scn/issues/221

### PR DESCRIPTION
**Pull request by javorszky**
Closes #143 

~~This changeset depends on https://github.com/suborbital/vektor/pull/78 being available. To get around the requirements, you can add the following line to atmo's `go.mod` file temporarily~~

> ```go
> replace github.com/suborbital/vektor => ../vektor
> ```

~~where `../vektor` points to a directory with the vektor codebase that's checked out on the branch for the pr 78.~~

This has been fixed in the go.mod file now.

All the tests pass locally, however I do need some help manually testing this. I *think* this should work? But I'm missing context on how to assemble atmo in proxy mode and pass the necessary configurations.

Speaking of configurations, the following environment variables have been introduced:

```go
type TracerConfig struct {
	Endpoint    string  `env:"ATMO_TRACER_ENDPOINT"`
	ServiceName string  `env:"ATMO_TRACER_SERVICENAME,default=atmo"`
	Probability float64 `env:"ATMO_TRACER_PROBABILITY,default=0.5"`
	APIKey      string  `env:"ATMO_TRACER_HONEYCOMB_APIKEY"`
	Dataset     string  `env:"ATMO_TRACER_HONEYCOMB_DATASET"`
}
```

These correspond to info needed by the otel tracer (service name, probability), and honeycomb (endpoint, apikey, dataset).

I worked off of the [quickstart documentation for go on honeycomb's pages](https://docs.honeycomb.io/getting-started/quickstart/#step-3-instrument-your-application-to-send-telemetry-data-to-honeycomb).

## How to test

The easiest way is to use this PR / branch from the sat repo: https://github.com/suborbital/sat/pull/42

It has a lot of great convenience features. With that, follow these steps:

1. check out atmo on this branch, and build the atmo-proxy with `make atmo/proxy/install`
2. [optional] check out the sat repo on the branch in the PR I linked above, this is useful for later
3. check that the sat repo also has an empty `traces.json` file in it. One of the exporters configured in the `oltp-config.yaml` file writes to that.
4. log in to honeycomb, and grab your API key
5. set the following environment variables:
    ```
    export ATMO_TRACER_HONEYCOMB_API_KEY="<your api key from above>"
    export ATMO_TRACER_HONEYCOMB_API_ENDPOINT="api.honeycomb.io:443"
    export ATMO_TRACER_HONEYCOMB_DATASET="<name it something, hc uses this in the web ui>"
    export ATMO_TRACER_TYPE="honeycomb" // this is either this, or "collector"
    export ATMO_TRACER_PROBABILITY=1
    ```
6. Start with the honeycomb test
    1. make sure tracer type is `"honeycomb"`
    2. make sure dataset is unique, and ends with `-hc` for easy identification
    3. Start constd with `make constd/metal/otel/hc`. It should start up correctly
    4. In the logs you should see log entries with text `configuring honeycomb exporter for tracing`, `created honeycomb trace exporter`, and `finished setting up tracer [honeycomb] with a trace probability of [1.0000]`
    5. using your choice of sending http requests, send a bunch to `POST localhost:8080/hello` with some body. Do a few (10-20) which should fill up the batcher
    6. log in to honeycomb, click on Datasets on the left, and you should see your new bucket with data in it
    7. stop constd on the conclusion of a successful test
7. Test the collector running beside constd
    1. change the dataset to something else, for example ending in `-collector`
    2. in the sat folder checked out to the PR I linked above, start up the collector with `docker compose up`. It should start up correctly, and just hang there taking up your terminal. This is expected.
    3. change the tracer type to "collector"
    4. double check that the collector docker service is still running
    5. start constd with `make constd/metal/otel/collector`. In should start up correctly
    6. in the logs you should see logs with texts `configuring collector exporter for tracing`, `created collector trace exporter`, and `finished setting up tracer [collector] with a trace probability of [1.0000]`
    7. using your choice of method, send a bunch of requests at `POST localhost:8080/hello` with some body. Do it 10-20 times for the batcher to kick in
    8. afterwards the size of the `traces.json` file should be more than 0B, and the content should be lines of json as logs
    9. honeycomb should also have these data, and they should be in a new bucket ending in `-collector`
    11. stop constd
    12. stop the collector docker
8. test the `none` tracer type
    1. change the tracer type to either an empty string `""`, or `"none"`
    2. change the dataset to end in `-none`
    3. truncate `traces.json` file
    4. start up constd with `make constd/metal/otel/collector`, or `make constd/metal/otel/hc`, it shouldn't matter, both should result in the same outcome; it should start up without issues
    5. in the startup logs you should see entries with texts `finished setting up noop tracer` and nothing else related to traces
    6. send a bunch of requests to `POST localhost:8080/hello` with whatever body
    7. honeycomb should not have a new dataset, and the traces.json file should remain empty
    8. stop constd
9. test the unknown tracer type
    1. set the tracer type to anything besides an empty string, `none`, `collector`, or `honeycomb`
    2. set the dataset to end in `-unknown`
    3. start up constd with either of the otel commands. It should start up correctly
    4. in the logs you should see two entries:
        1. one warning with `unrecognised tracer type configuration [<your unknown type>]. Defaulting to no tracer`
        2. the usual `finished setting up noop tracer`
    5. send a bunch of (10-20) requests to `POST localhost:8080/hello` with whatever body
    6. honeycomb should not have data with that dataset, and the traces.json file should remain empty
    7. stop constd

### CODE REVIEW

**Comment by javorszky** [comment]
It fails because the vektor dependency needs to be updated, but that's currently sitting in a PR

**Comment by javorszky** [comment]
This now works if the collector is on localhost:4317, which is the default for the opentel collector.

I was thinking that, as this is an open source codebase, the configuration should be expanded such that:

```yaml
telemetryProvider: none | collector | honeycomb | jaeger | zipkin
telemetrySettings:
    collector:
        # collector specific settings go here
    honeycomb:
        # honeycomb specific settings go here
    jaeger:
        # jaeger specific settings go here
    zipkin:
        # zipkin specific settings go here
```
And the options struct be modified similary.

The idea being that rather than infer and decide, default to "none" (in which case it's a nooptracer), and have the users configure the tracer as they see fit. This way there's precision in what should be done rather than weird unexpected outcomes.

File: atmo/coordinator/coordinator_tracing.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	StatusCode int
+}
+
+func (c *Coordinator) openTelemetryHandler() vk.Middleware {
+	return func(r *http.Request, ctx *vk.Ctx) error {
+		tracedCtx, span := otel.GetTracerProvider().Tracer("").Start(ctx.Context, "coordinator.openTelemetryHandler")
+
+		ctx.Context = tracedCtx
+
+		v := Values{
```
Comment (cohix) [inline atmo/coordinator/coordinator_tracing.go:28]: Are we going to add the `request ID` to this?

File: atmo/init_proxy.go [hunk 2]
```diff
@@ -2,8 +2,144 @@
...
+
+func honeycombExporterOptions(config *options.HoneycombConfig) ([]otlptracegrpc.Option, error) {
+	if config == nil {
+		return nil, errors.New("empty honeycomb tracer configuration")
+	}
+
+	return []otlptracegrpc.Option{
+		otlptracegrpc.WithEndpoint(config.Endpoint),
+		otlptracegrpc.WithHeaders(map[string]string{
+			"x-honeycomb-team":    config.APIKey,
```
Comment (cohix) [inline atmo/init_proxy.go:140]: Odd that the APIKey goes into the `team` header? 🤔

Comment (javorszky) [reply atmo/init_proxy.go:140]: It is, but it's what they say we should do: https://docs.honeycomb.io/getting-data-in/opentelemetry/go/

File: README.md [hunk 3]
```diff
@@ -43,3 +43,24 @@ Visit [the Suborbital website](https://suborbital.dev) to sign up for email upda
 Atmo is built atop [Vektor](https://github.com/suborbital/vektor), [Grav](https://github.com/suborbital/grav), and [Reactr](https://github.com/suborbital/reactr).
 
 Copyright Suborbital contributors 2021.
+
+## Configuration
```
[the remark this answers is not in the document]
Comment (javorszky) [reply README.md:47]: I can see that, and I'm happy to move it across. I wonder whether a link to the Atmo Guide is a strong enough signifier that the entire documentation lives there, including technical details though.

[the remark this answers is not in the document]
Comment (javorszky) [reply README.md:47]: Moved it to here: https://github.com/suborbital/docs/pull/61

[the remark this answers is not in the document]
Comment (javorszky) [reply README.md]: And removed the commit for this from here

**Review by cohix — APPROVED** [review]
Tested and works with Honeycomb and local collector!

**Pull request merged**